cherab/tools/spectroscopy/spectrometer.py

```diff
@@ -0,0 +1,298 @@
+
+# Copyright 2016-2021 Euratom
+# Copyright 2016-2021 United Kingdom Atomic Energy Authority
+# Copyright 2016-2021 Centro de Investigaciones Energéticas, Medioambientales y Tecnológicas
+#
+# Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the
+# European Commission - subsequent versions of the EUPL (the "Licence");
+# You may not use this work except in compliance with the Licence.
+# You may obtain a copy of the Licence at:
+#
+# https://joinup.ec.europa.eu/software/page/eupl5
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied.
+#
+# See the Licence for the specific language governing permissions and limitations
+# under the Licence.
+
+import numpy as np
+from raysect.optical.observer import SpectralRadiancePipeline0D
+
+from .instrument import SpectroscopicInstrument
+
+
+class Spectrometer(SpectroscopicInstrument):
+    """
+    Spectrometer base class.
+    This is an abstract class.
+
+    :param int spectral_bins: The number of spectral samples over the wavelength range.
+    :param float reference_wavelength: Wavelength (in nm) corresponding to
+                                       the centre of reference bin.
+    :param int reference_bin: Reference bin index. Can be negative to specify the offset.
+                              Default is None (spectral_bins // 2).
+    :param str name: Spectrometer name.
+    """
+
+    def __init__(self, spectral_bins, reference_wavelength, reference_bin=None, name=''):
```

Comment:
What is meant by the reference wavelength and/or reference bin and does it have to be specified here in the abstract class? Wouldn't it be more general to specify only min_wavelength, max_wavelength and resolution?

 Also, when having these classes, it would be great if there was an attribute for specifying a calibration function (pixel->wavelength), which might be FunctionxD, depending on the detector. I understand this would not be useful for the observations, because we can have only a constant step per bin, but it could be useful for postprocessing and comparison of measured and modelled data. It would be aiming more at people operating a diagnostic, rather than people getting already wavelength values for bins. I would definitely use such tool when working on "raw" data obtained from spectrometer,. What do you think?

Also, some imaging spectrometers can accommodate more spectra (separated in 1 or 2D on a camera chip), is this doable with the suggested architecture?

Reply:
The reference wavelength and reference pixel make sense to JET and reflect the way the spectrometer settings are stored in JPF. So, I agree that these should be removed.

Regarding the pixel->wavelength calibration, I think that the `Spectrometer` can be initialized with a sequence of calibration arrays (one array per accommodated spectrum) containing the wavelengths of the pixel borders. Namely, the values w<sup>k</sup><sub>i</sub> and w<sup>k</sup><sub>i+1</sub> define the spectral range of the pixel p<sub>i</sub> of the k-th spectrum. After the spectrum is ray-traced, we can calibrate it with: 
```python
calibrated_spectra = spectrometer.calibrate(spectrum)
```
where `calibrate(spectrum)` will do the integration:
```python
    def calibrate(self, spectrum):
        if not isinstance(spectrum, Spectrum):
            raise TypeError('Argument spectrum must be a Spectrum instance.')
        calibrated_spectra = []
        for cal_array in self.calibration_arrays:
            calibrated_spectrum = np.zeros(cal_array.size - 1)
            for i in range(cal_array.size - 1):
                calibrated_spectrum[i] = spectrum.integrate(cal_array[i], cal_array[i + 1])
            calibrated_spectra.append(calibrated_spectrum)

        return calibrated_spectra
```
With this approach we don't care if the spectrum is mapped on the chip in 1D or 2D. Of course, the machine specific spectrometers can be initialized with the known function to calculate these calibration arrays.

The `CzernyTurnerSpectrometer` will require a sequence of `(min_wavelength, num_pixels)` values instead of the calibration arrays to initialize. The calibration arrays will be calculated based on the specified parameters of the diffraction scheme.

One can control the number of bins of ray-traced spectrum through the `min_bins_per_pixel` property.

@Mateasek, what do you think of this solution?

Reply:
I've reworked the spectrometers according to the proposed scheme. I think they are much more useful now.